core/src/main/java/io/parsingdata/metal/data/ByteStreamSource.java

```diff
@@ -28,18 +29,22 @@
 
     public final ByteStream input;
 
-    public ByteStreamSource(final ByteStream input) {
+    ByteStreamSource(final ByteStream input) {
         this.input = checkNotNull(input, "input");
     }
 
     @Override
-    protected byte[] getData(final long offset, final BigInteger length) throws IOException {
-        if (!isAvailable(offset, length)) { throw new IOException("Data to read is not available ([offset=" + offset + ";length=" + length + ";source=" + this + ")."); }
-        return input.read(offset, length.intValue());
+    protected byte[] getData(final long offset, final BigInteger length) {
```

Comment:
Just for information, as you want this Metal implementation to be functional/immutable and strive for correctness, should we then not check these arguments as well, in order to more easily spot possible mistake (e.g., if ever a negative length is passed here?)